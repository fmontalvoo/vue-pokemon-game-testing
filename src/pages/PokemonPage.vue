<template>
    <span v-if="!pokemon" class="loader"></span>

    <div v-else-if="pokemon">
        <h2>Â¿Qui&eacute;n es este pok&eacute;mon?</h2>
        <PokemonPicture :pokemon-id="pokemon.id" :show-pokemon="showPokemon" />
        <PokemonOptions :pokemon-options="pokemons" @select-pokemon="checkAnswer" />

        <div class="fade-in" v-show="showPokemon">
            <h2>!Es {{ capitalizedName(pokemon.name) }}!</h2>
            <button @click="resetGame">Jugar de nuevo</button>
        </div>

    </div>

    <h3 v-else>
        Algo salio mal!!!
    </h3>
</template>

<script>
import getPokemonOptions, { getRandomInt, capitalize } from '@/utils/utils.js'

import PokemonPicture from '@/components/PokemonPicture.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'

export default {
    components: {
        PokemonPicture,
        PokemonOptions,
    },
    data() {
        return {
            pokemons: [],
            pokemon: null,
            showPokemon: false,
        }
    },
    mounted() {
        this.getPokemos()
    },
    methods: {
        async getPokemos() {
            this.pokemons = await getPokemonOptions()
            this.pokemon = this.pokemons.at(getRandomInt(0, 3))
        },
        checkAnswer(pokemonId) {
            console.log(pokemonId)
            this.showPokemon = true
        },
        resetGame() {
            this.pokemons = []
            this.pokemon = null
            this.showPokemon = false
            this.getPokemos()
        },
        capitalizedName(str) {
            return capitalize(str)
        }
    }
}
</script>

<style>
/* https://cssloaders.github.io/ */
.loader {
    width: 48px;
    height: 48px;
    border: 3px dotted crimson;
    border-style: solid solid dotted dotted;
    border-radius: 50%;
    display: inline-block;
    position: relative;
    box-sizing: border-box;
    animation: rotation 2s linear infinite;
}

.loader::after {
    content: '';
    box-sizing: border-box;
    position: absolute;
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    margin: auto;
    border: 3px dotted cadetblue;
    border-style: solid solid dotted;
    width: 24px;
    height: 24px;
    border-radius: 50%;
    animation: rotationBack 1s linear infinite;
    transform-origin: center center;
}

@keyframes rotation {
    0% {
        transform: rotate(0deg);
    }

    100% {
        transform: rotate(360deg);
    }
}

@keyframes rotationBack {
    0% {
        transform: rotate(0deg);
    }

    100% {
        transform: rotate(-360deg);
    }
}
</style>