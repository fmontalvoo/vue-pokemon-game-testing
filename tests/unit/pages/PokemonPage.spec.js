import { shallowMount, mount } from '@vue/test-utils'

import PokemonPage from '@/pages/PokemonPage.vue'

describe('PokemonPage', () => {
    let wrapper;
    const pokemons = [
        { id: 1, name: 'bulbasaur' },
        { id: 4, name: 'charmander' },
        { id: 7, name: 'squirtle' },
        { id: 25, name: 'pikachu' },
    ];

    test('Debe llamar al metodo getPokemos() al instanciar el componente', () => {
        const getPokemosSpy = jest.spyOn(PokemonPage.methods, 'getPokemos')
        const wrapper = shallowMount(PokemonPage)

        expect(getPokemosSpy).toHaveBeenCalled()
    })

    test('Debe hacer match con el snapshot al cargar la informacion', () => {
        wrapper = mount(PokemonPage, {
            data() {
                return {
                    pokemons: pokemons,
                    pokemon: pokemons[2],
                    showPokemon: false,
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    })

    test('Debe mostrar los componentes hijos', () => {
        wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemons: pokemons,
                    pokemon: pokemons[2],
                    showPokemon: false,
                }
            }
        })

        const PokemonPicture = wrapper.find('pokemon-picture-stub')
        const PokemonOptions = wrapper.find('pokemon-options-stub')

        // console.info(PokemonPicture.html())
        // console.info(PokemonOptions.html())

        expect(PokemonPicture.exists()).toBeTruthy()
        expect(PokemonOptions.exists()).toBeTruthy()

        expect(PokemonOptions.attributes('pokemonoptions')).toBeTruthy()
        expect(PokemonPicture.attributes('pokemonid')).toEqual(pokemons[2].id.toString())
    })

    test('Debe llamar al metodo checkAnswer()', async () => {
        wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemons: pokemons,
                    pokemon: pokemons[3],
                    showPokemon: false,
                }
            }
        })

        const checkAnswerSpy = jest.spyOn(wrapper.vm, 'checkAnswer')

        await wrapper.vm.checkAnswer(25)

        expect(wrapper.vm.showPokemon).toBe(true)
        expect(checkAnswerSpy).toHaveBeenCalledWith(25)

        const h2 = wrapper.find('[data-testid="pokemon-name"]')

        expect(h2.text().toLowerCase()).toContain(pokemons[3].name)
    })
})