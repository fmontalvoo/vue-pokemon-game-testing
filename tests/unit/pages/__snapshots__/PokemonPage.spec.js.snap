// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Debe hacer match con el snapshot al cargar la informacion 1`] = `
<div>
  <h2>¿Quién es este pokémon?</h2>
  <div class="pokemon-container"><img class="hidden-pokemon" src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/7.svg" alt="Pokemon"></div>
  <div class="options-container">
    <ul>
      <li>Bulbasaur</li>
      <li>Charmander</li>
      <li>Squirtle</li>
      <li>Pikachu</li>
    </ul>
  </div>
  <div class="fade-in" style="display: none;">
    <h2 data-testid="pokemon-name">!Es Squirtle!</h2><button>Jugar de nuevo</button>
  </div>
</div>
`;
